File: test/src/main/java/org/sqlunet/browser/BaseWaitUntil.java
/*
 * Copyright (c) 2023. Bernard Bou
 */

package org.sqlunet.browser;

import android.view.View;

import org.hamcrest.Matcher;

import java.lang.reflect.Field;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.test.espresso.IdlingResource;
import androidx.test.espresso.ViewFinder;
import androidx.test.espresso.ViewInteraction;

import static androidx.test.espresso.Espresso.onView;

public abstract class BaseWaitUntil implements IdlingResource
{
	protected final Matcher<View> viewMatcher;

	protected ResourceCallback resourceCallback;

	public BaseWaitUntil(final Matcher<View> viewMatcher)
	{
		this.viewMatcher = viewMatcher;
	}

	@Override
	public void registerIdleTransitionCallback(final ResourceCallback resourceCallback)
	{
		this.resourceCallback = resourceCallback;
	}

	@Nullable
	protected static View getView(@NonNull Matcher<View> viewMatcher)
	{
		try
		{
			final ViewInteraction viewInteraction = onView(viewMatcher);
			// private in ViewInteraction
			final Field finderField = viewInteraction.getClass().getDeclaredField("viewFinder");
			finderField.setAccessible(true);
			final ViewFinder finder = (ViewFinder) finderField.get(viewInteraction);
			assert finder != null;
			return finder.getView();
		}
		catch (Exception e)
		{
			return null;
		}
	}
}

File: test/src/main/java/org/sqlunet/browser/BaseWaitUntil.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.browser

import android.view.View
import androidx.test.espresso.Espresso
import androidx.test.espresso.IdlingResource
import androidx.test.espresso.IdlingResource.ResourceCallback
import androidx.test.espresso.ViewFinder
import org.hamcrest.Matcher

abstract class BaseWaitUntil(@JvmField protected val viewMatcher: Matcher<View?>) : IdlingResource {

    @JvmField
    protected var resourceCallback: ResourceCallback? = null

    override fun registerIdleTransitionCallback(resourceCallback: ResourceCallback) {
        this.resourceCallback = resourceCallback
    }

    companion object {

        @JvmStatic
        protected fun getView(viewMatcher: Matcher<View?>): View? {
            return try {
                val viewInteraction = Espresso.onView(viewMatcher)
                // private in ViewInteraction
                val finderField = viewInteraction.javaClass.getDeclaredField("viewFinder")
                finderField.isAccessible = true
                val finder = (finderField[viewInteraction] as ViewFinder)
                finder.view
            } catch (e: Exception) {
                null
            }
        }
    }
}
